### PR TITLE
File Operator Fixes

### LINKED ISSUE #202: Refactor file operator code and complete git init level
level `intro/init` is used to teach someone to use `git init`, but for them to do that, the current directory should not be a git repo. In order to do that,we would need to add several layers of convolution to get the file operators but instead, let's just refactor

Task:
 - Remove ownership of file operators from the GitGud object
 - Never pass file operators, instead get it from `operations.get_operator()`  (Never re-initialize)
 - Make sure `git init` acts as if we aren't in a git repo when the level is initialized and the shows up as it would normally after running `git init`
 - Refactor existing code that uses file operators to reflect changes
 - Fix anything else that's relevant to this and you think makes sense. You're at liberty to change the structure of the code here if you think it'd help

Base changes on and merge into https://github.com/benthayer/git-gud/tree/intro

### PR DESCRIPTION
**Pull request by sahansk2**
Fixes #202 
The most notable difference thus far is the use of the builtin `pickle` module to reuse the same file operator throughout.

(This PR is still in progress).
Tasks from #202 :
- [x] Remove ownership of file operators from the GitGud object
- [x] Never pass file operators, instead get it from operations.get_operator() (Never re-initialize)
- [x] Make sure git init acts as if we aren't in a git repo when the level is initialized and the shows up as it would normally after running git init
- [x] Refactor existing code that uses file operators to reflect changes
- [ ] Fix anything else that's relevant to this and you think makes sense. You're at liberty to change the structure of the code here if you think it'd help

### CODE REVIEW

**Comment by sahansk2** [comment]
Oops, I forgot to edit the description after my constant rebasing. The GitGud object no longer has a file_operator attribute. There isn't a single function that takes a file_operator argument anymore.

File: gitgud/__main__.py [hunk 1]
```diff
@@ -203,30 +201,39 @@ def __init__(self):
...
-                level_name = self.file_operator.read_level_file()
+                level_name = operations.get_operator().read_level_file()
                 raise InitializationError(
                         'Currently loaded level does not exist: "{}"'
                         .format(level_name))
 
     def load_level(self, level):
         # Clear remotes
+        file_operator = operations.get_operator()
         self.assert_initialized(skip_level_check=True)
```
Comment (benthayer) [inline gitgud/__main__.py:222]: `assert_initialized` should always be the first line

File: gitgud/operations.py [hunk 2]
```diff
@@ -91,7 +106,7 @@ def create_tree(self, commits, head):
             if parents:
                 # TODO GitPython detach head
                 self.repo.git.checkout(parents[0])
-            if len(parents) < 2:
+            if len(parents) < 22:
```
Comment (benthayer) [inline gitgud/operations.py:109]: Typo?

File: gitgud/skills/intro/__init__.py [hunk 3]
```diff
@@ -73,13 +76,14 @@ def status(self):
         else:
             self.display_message("status-norepo.txt")
 
-    def _test(self, file_operator):
+    def _test(self):
         # Check if we are in a git repo
+
```
Comment (benthayer) [inline gitgud/skills/intro/__init__.py:81]: Do we need the extra line here?

File: gitgud/skills/intro/__init__.py [hunk 4]
```diff
@@ -73,13 +76,14 @@ def status(self):
         else:
             self.display_message("status-norepo.txt")
```
Comment (benthayer) [inline gitgud/skills/intro/__init__.py:77]: `git gud status` doesn't produce correct output. It always thinks we're in a repo

File: gitgud/skills/intro/__init__.py (hunk 3 above)
Comment (sahansk2) [reply gitgud/skills/intro/__init__.py:81]: Yeah, flake8 complains otherwise.

**Comment by sahansk2** [comment]
> do we really want to write changes every time? Does it matter that we have it saved on disk? What was your thought process?

In general, I don't like writing to disk unless absolutely necessary, since it generally feels unsafe, unsure, and otherwise clunky. The main reason I went about using `pickle` was because of the point of "never reinitializing". There weren't that many ways to maintain a sort of "persistent" state between Python script invocations to ensure `Operator` never reinitializes, since runtime stuff is discarded after the script is done. One solution was using the above: by using the `pickle` module, we just store the state of the file_operator in a file. The caveat there is that we have to access the operator state through the `os.path` module everytime. I admit, this is kind of rough, and I'm not a fan of it either.

Another option I was looking at was `reload` in `importlib`. But that seems like there has to be a continuously running script to keep track of the state there, so I decided against that. https://stackoverflow.com/questions/6687660/keep-persistent-variables-in-memory-between-runs-of-python-script

I could make a fix where the state of an `Operator` instance is based on the state of the `.git/` directory though. If `.git/gud` exists, git gud is initialized. if `.git/HEAD` also exists, there's a gitgud-facilitated repo. If there's only a `.git/HEAD` but no `.git/gud`, then git gud isn't initialized, and we're currently in an active repo. If `.git/gud` exists but there is no `.git/HEAD`, then we're in an uninitialized level. And it goes on.

> What do you think about officially making Operator static, since that's how we're currently using it? 

I'm cool with this if we keep track of state through the `.git/` directory.

File: gitgud/operations.py (hunk 2 above)
Comment (sahansk2) [reply gitgud/operations.py:109]: Oops, yes it is. :sweat_smile:

**Comment by sahansk2** [comment]
Alright, I'm closing this issue and making a new PR based on `intro`. There's so much code that used `pickle`. Now that I know that we wanted to reuse file_operator within a single execution, we don't need any of that `pickle` business. Not to mention that the `pickle` commits are just going to be fluff in the history.

I'll probably cherrypick commit be8df826a18362bbe15f311308fe778591d554c3 though, since it's got useful `Init` level code.

**Comment by benthayer** [comment]
I thought I removed most of the pickle code?
Rebasing and force pushing here would be fine, it might be nice to preserve the single pull request. It seems like this is close to being done. Feel free to change things though

**Comment by sahansk2** [comment]
You know what, sure, I'll keep developing on this branch ¯\\\_(ツ)_/¯

**Comment by sahansk2** [comment]
Certain levels break on their loading after immediately loading level `0-3`. It's never permanent since trying to load the level again works. Here are the levels that don't load properly after loading `0-3`:
- All levels in skills 2, rampup:
   - 1: detaching
   - 2: relrefs1
   - 3: relrefs2
   - 4: reversing
- Skill 4 (extras) level 1 (octopus)

**Comment by sahansk2** [comment]
For reference: This is the loading error that I was talking about:
```
sahan@sahan-Lenovo-YOGA-C930-ubuntu:~/Documents/osai/ggtest$ git gud load 0 3
Git is a version control system. This means that it stores and tracks the
changes you make. Before we get to the business of making commits and branches
and doing other fun things, we'll need a place to store all of it. That place
is known as a "Git Repo".

To see the goal for this level, type:
    $ git gud goal
sahan@sahan-Lenovo-YOGA-C930-ubuntu:~/Documents/osai/ggtest$ git gud load 2 2
Traceback (most recent call last):
  File "/home/sahan/.local/bin/git-gud", line 11, in <module>
    load_entry_point('git-gud', 'console_scripts', 'git-gud')()
  File "/home/sahan/Documents/osai/git-gud/gitgud/__main__.py", line 533, in main
    GitGud().parse()
  File "/home/sahan/Documents/osai/git-gud/gitgud/__main__.py", line 527, in parse
    self.command_dict[args.command](args)
  File "/home/sahan/Documents/osai/git-gud/gitgud/__main__.py", line 466, in handle_load
    self.load_level(level)
  File "/home/sahan/Documents/osai/git-gud/gitgud/__main__.py", line 230, in load_level
    level.setup()
  File "/home/sahan/Documents/osai/git-gud/gitgud/skills/level_builder.py", line 42, in setup
    self._setup()
  File "/home/sahan/Documents/osai/git-gud/gitgud/skills/level_builder.py", line 101, in _setup
    file_operator.create_tree(commits, head)
  File "/home/sahan/Documents/osai/git-gud/gitgud/operations.py", line 150, in create_tree
    self.repo.create_head(branch, self.repo.head.commit)
  File "/home/sahan/.local/lib/python3.6/site-packages/git/repo/base.py", line 386, in create_head
    return Head.create(self, path, commit, force, logmsg)
  File "/home/sahan/.local/lib/python3.6/site-packages/git/refs/symbolic.py", line 543, in create
    return cls._create(repo, path, cls._resolve_ref_on_create, reference, force, logmsg)
  File "/home/sahan/.local/lib/python3.6/site-packages/git/refs/symbolic.py", line 506, in _create
    (full_ref_path, existing_data, target_data))
OSError: Reference at 'refs/heads/master' does already exist, pointing to '19cedd17ee6b851f6982c72f47e8d57501a9abc9', requested was '27e198341c014e09e679f4a02a9f73c8643c013e'
```

**Comment by sahansk2** [comment]
I'm not entirely sure what's going on with pytest right now -- it's working perfectly fine on my local machine, except for the loading weirdness. Feel free to take a look, @benthayer

**Pull request closed without merging**